pkg/remote/remote.go

```diff
@@ -0,0 +1,320 @@
+	DocRev        string `json:"_rev,omitempty"`
+	RemoteDoctype string `json:"doctype"`
+	Verb          string `json:"verb"`
+	URL           string `json:"url"`
+	ResponseCode  int    `json:"response_code"`
+	ContentType   string `json:"content_type"`
+}
+
+// ID is used to implement the couchdb.Doc interface
+func (r *Request) ID() string { return r.DocID }
+
+// Rev is used to implement the couchdb.Doc interface
+func (r *Request) Rev() string { return r.DocRev }
+
+// SetID is used to implement the couchdb.Doc interface
+func (r *Request) SetID(id string) { r.DocID = id }
+
+// SetRev is used to implement the couchdb.Doc interface
+func (r *Request) SetRev(rev string) { r.DocRev = rev }
+
+// DocType implements couchdb.Doc
+func (r *Request) DocType() string { return consts.RemoteRequests }
+
+// Clone implements couchdb.Doc
+func (r *Request) Clone() couchdb.Doc { cloned := *r; return &cloned }
+
+// Remote is the struct used to call a remote website for a doctype
+type Remote struct {
+	Verb    string
+	URL     *url.URL
+	Headers map[string]string
+	Body    string
+}
+
+var log = logger.WithNamespace("remote")
+
+// ParseRawRequest takes a string and parse it as a remote struct.
+// First line is verb and URL.
+// Then, we have the headers.
+// And for a POST, we have a blank line, and then the body.
+func ParseRawRequest(doctype, raw string) (*Remote, error) {
+	lines := strings.Split(raw, "\n")
+	parts := strings.Split(lines[0], " ")
+	if len(parts) != 2 {
+		log.Infof("%s cannot be used as a remote doctype", doctype)
+		return nil, ErrInvalidRequest
+	}
+	var remote Remote
+	remote.Verb = parts[0]
+	if remote.Verb != echo.GET && remote.Verb != echo.POST {
+		log.Infof("Invalid verb for remote doctype %s: %s", doctype, remote.Verb)
+		return nil, ErrInvalidRequest
+	}
+	u, err := url.Parse(parts[1])
+	if err != nil {
+		log.Infof("Invalid URL for remote doctype %s: %s", doctype, parts[1])
+		return nil, ErrInvalidRequest
+	}
+	if u.Scheme != "https" && u.Scheme != "http" {
+		log.Infof("Invalid scheme for remote doctype %s: %s", doctype, u.Scheme)
+		return nil, ErrInvalidRequest
+	}
+	remote.URL = u
+	remote.Headers = make(map[string]string)
+	for i, line := range lines[1:] {
+		if line == "" {
+			if remote.Verb == echo.GET {
+				continue
+			}
+			remote.Body = strings.Join(lines[i+2:], "\n")
+			break
+		}
+		parts = strings.SplitN(line, ":", 2)
+		if len(parts) != 2 {
+			log.Infof("Invalid header for remote doctype %s: %s", doctype, line)
+			return nil, ErrInvalidRequest
+		}
+		remote.Headers[parts[0]] = strings.TrimSpace(parts[1])
+	}
+	return &remote, nil
+}
+
+// Find finds the request defined for the given doctype
+func Find(ins *instance.Instance, doctype string) (*Remote, error) {
+	var raw string
+
+	if config.GetConfig().Doctypes == "" {
+		dt := Doctype{
+			DocID: consts.Doctypes + "/" + doctype,
+		}
+		err := couchdb.GetDoc(ins, consts.Doctypes, dt.DocID, &dt)
+		if err != nil {
```

Comment:
Yes, it was my intention to use github even for couchdb internal errors. Do you see some cons to that?